bundles/org.openhab.binding.e3dc/src/main/java/org/openhab/binding/e3dc/internal/modbus/ModbusCallback.java

```diff
@@ -0,0 +1,142 @@
+
+import java.util.Arrays;
+import java.util.Iterator;
+import java.util.Optional;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.openhab.binding.e3dc.internal.dto.EmergencyBlock;
+import org.openhab.binding.e3dc.internal.dto.InfoBlock;
+import org.openhab.binding.e3dc.internal.dto.PowerBlock;
+import org.openhab.binding.e3dc.internal.dto.StringBlock;
+import org.openhab.binding.e3dc.internal.dto.WallboxArray;
+import org.openhab.binding.e3dc.internal.modbus.Data.DataType;
+import org.openhab.io.transport.modbus.AsyncModbusFailure;
+import org.openhab.io.transport.modbus.AsyncModbusReadResult;
+import org.openhab.io.transport.modbus.ModbusFailureCallback;
+import org.openhab.io.transport.modbus.ModbusReadCallback;
+import org.openhab.io.transport.modbus.ModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.ModbusRegister;
+import org.openhab.io.transport.modbus.ModbusRegisterArray;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link InfoBlockCallback} class receives callbacks from modbus poller
+ *
+ * @author Bernd Weymann - Initial contribution
+ */
+@NonNullByDefault
+public class ModbusCallback extends ModbusDataProvider
+        implements ModbusReadCallback, ModbusFailureCallback<ModbusReadRequestBlueprint> {
+    private final Logger logger = LoggerFactory.getLogger(ModbusCallback.class);
+    private DataType callbackType;
+    private byte[] bArray;
+    private int size;
+    private int counter = 0;
+    private long maxDuration = Long.MIN_VALUE;
+    private long minDuration = Long.MAX_VALUE;
+    private long avgDuration = 0;
+
+    public ModbusCallback(DataType type) {
+        callbackType = type;
+        if (type.equals(DataType.INFO)) {
+            size = INFO_REG_SIZE * 2;
+            bArray = new byte[size];
+        } else {
+            size = (REGISTER_LENGTH - INFO_REG_SIZE) * 2;
+            bArray = new byte[size];
+        }
+    }
+
+    @Override
+    public void handle(AsyncModbusReadResult result) {
+        byte[] newArray = new byte[size];
+        long startTime = System.currentTimeMillis();
+        Optional<ModbusRegisterArray> opt = result.getRegisters();
+        ModbusRegisterArray registers = opt.get();
+        Iterator<ModbusRegister> iter = registers.iterator();
+        int i = 0;
+        while (iter.hasNext()) {
+            ModbusRegister reg = iter.next();
+            System.arraycopy(reg.getBytes(), 0, newArray, i, 2);
+            i += 2;
+        }
+        setArray(newArray);
+
+        long duration = System.currentTimeMillis() - startTime;
+        avgDuration += duration;
+        minDuration = Math.min(minDuration, duration);
+        maxDuration = Math.max(maxDuration, duration);
+        counter++;
+        if (counter % 100 == 0) {
+            logger.debug("Min {} Max {} Avg {}", minDuration, maxDuration, avgDuration / 30);
```

Comment:
average calculated wrong? Should divide by 100?

Reply:
Sure, values changed while measuring and I forgot to update both values